server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -360,7 +372,7 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
 
     private final AtomicLong idGenerator = new AtomicLong();
 
-    private final Map<Long, ReaderContext> activeReaders = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private final Map<ReaderContextId, ReaderContext> activeReaders = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
```

Comment:
Currently we keep track of the readers in this map only by an auto-increment long counter, which is also encoded in the PIT id to lookup the right ReaderContext. We cannot simply move the ReaderContext to another node using only that id now because of conflicts with existing contexts on that node. But we also cannot change it because its encoded in the PIT Id (via the ShardSearchContextIds). I changed the key to a new joint key that also uses the original sessionId encoded in the PIT for retrieval on the new node and to avoid id collisions.